File: static/js/theme.js
// Shared theme system — loaded on every page from base.html.
//
// Preference model: the user picks one of 'light' | 'dark' | 'system'.
// 'system' resolves to light/dark from the OS and reacts live to OS changes.
// The preference is persisted two ways:
//   - localStorage ('syncchat:theme')   — anonymous users / before login / instant
//   - database (User.theme, via the API) — logged-in users, survives devices
// The inline no-flash script in base.html already applied the initial value
// before this file runs; here we just keep it in sync and expose the API.
(function () {
  const LS_KEY = "syncchat:theme";
  const VALID = ["light", "dark", "system"];
  const root = document.documentElement;

  /* ---------- Resolution helpers ---------- */
  function osPrefersDark() {
    return (
      window.matchMedia &&
      window.matchMedia("(prefers-color-scheme: dark)").matches
    );
  }

  function resolve(pref) {
    if (pref === "system") return osPrefersDark() ? "dark" : "light";
    return pref === "dark" ? "dark" : "light";
  }

  // The inline script stashed the chosen preference here; fall back to
  // localStorage, then to 'system'.
  function getPreference() {
    let pref = root.getAttribute("data-theme-pref");
    if (VALID.indexOf(pref) === -1) {
      try { pref = localStorage.getItem(LS_KEY); } catch (e) { pref = null; }
    }
    if (VALID.indexOf(pref) === -1) pref = "system";
    return pref;
  }

  function getResolved() {
    return root.dataset.theme || resolve(getPreference());
  }

  /* ---------- Smooth transitions ----------
     Bare CSS-variable swaps don't animate because variable changes aren't
     transitionable. So we briefly enable transitions on the common colour
     properties only during an explicit switch, then turn them off again so
     the initial paint and unrelated DOM mutations stay cheap. */
  let transitionTimer;

  function withTransitions(fn) {
    if (window.matchMedia && window.matchMedia("(prefers-reduced-motion: reduce)").matches) {
      fn();
      return;
    }
    root.classList.add("theme-transition");
    fn();
    clearTimeout(transitionTimer);
    transitionTimer = setTimeout(() => root.classList.remove("theme-transition"), 300);
  }

  /* ---------- Apply / persist ---------- */
  function applyResolved(resolved) {
    root.dataset.theme = resolved;
    // Let the page react (e.g. toggle button icons / aria labels).
    document.dispatchEvent(
      new CustomEvent("syncchat:themechange", {
        detail: {
          preference: getPreference(),
          resolved: resolved,
        },
      })
    );
  }

  function persist(pref) {
    try { localStorage.setItem(LS_KEY, pref); } catch (e) {}
    // Fire-and-forget DB save for logged-in users; anonymous users still
    // have localStorage so the next page load is correct.
    try {
      const token = (document.cookie.match(/csrftoken=([^;]+)/) || [])[1];
      if (token) {
        const body = new URLSearchParams({ theme: pref });
        fetch("/accounts/theme/", {
          method: "POST",
          headers: {
            "X-CSRFToken": token,
            "Content-Type": "application/x-www-form-urlencoded",
          },
          body: body.toString(),
        }).catch(() => {});
      }
    } catch (e) {}
  }

  // Set a preference (light|dark|system), resolve, apply, and persist.
  function setPreference(pref) {
    if (VALID.indexOf(pref) === -1) return;
    root.dataset.themePref = pref;
    persist(pref);
    const resolved = resolve(pref);
    withTransitions(() => applyResolved(resolved));
    return resolved;
  }

  // Simpler toggle used by the dashboard sun/moon button: flip light <-> dark.
  function toggle() {
    const next = getResolved() === "dark" ? "light" : "dark";
    return setPreference(next);
  }

  /* ---------- Live OS tracking for 'system' ---------- */
  function refreshIfSystem() {
    if (getPreference() !== "system") return;
    withTransitions(() => applyResolved(resolve("system")));
  }

  if (window.matchMedia) {
    const mql = window.matchMedia("(prefers-color-scheme: dark)");
    const handler = mql.addEventListener
      ? (e) => e.addEventListener("change", refreshIfSystem)
      : null;
    // addEventListener is the modern API; fall back silently otherwise.
    if (mql.addEventListener) {
      mql.addEventListener("change", refreshIfSystem);
    } else if (mql.addListener) {
      mql.addListener(refreshIfSystem);
    }
  }

  // Re-resolve on first load in case the inline script's server hint was stale
  // (e.g. OS preference changed since the page was rendered). Cheap and safe.
  if (getPreference() === "system") {
    applyResolved(resolve("system"));
  }

  window.SyncChatTheme = {
    getPreference,
    getResolved,
    resolve,
    setPreference,
    toggle,
    applyResolved,
  };
})();

File: core/theme.py
"""Theme preference resolution, shared by templates and views.

Resolution priority for a logged-in user:
  1. The user's saved DB preference (``theme`` field) — one of
     ``'light'``, ``'dark'``, ``'system'``.
  2. Fall back to ``'system'`` (the model default).

``'system'`` resolves to either ``'light'`` or ``'dark'`` based on the request
``prefers-color-scheme`` header when available, otherwise defaults to light.

The context processor exposes two values to every template:
  ``theme_pref``     — the raw preference ('light'|'dark'|'system')
  ``theme_resolved`` — the concrete value to apply ('light'|'dark'), so the
                      inline no-flash script can set ``data-theme`` without
                      waiting for matchMedia.
"""

THEME_LIGHT = "light"
THEME_DARK = "dark"
THEME_SYSTEM = "system"

VALID_THEMES = (THEME_LIGHT, THEME_DARK, THEME_SYSTEM)


def _prefers_dark(request):
    """Return True if the client signals a dark OS preference via the
    ``Sec-CH-Prefers-Color-Scheme`` or ``prefers-color-scheme`` header.

    Not all browsers send these headers, so an unknown preference falls back
    to light (also the model default). This is only needed to resolve
    ``'system'`` on the server; JS handles live OS changes after load.
    """
    if request is None:
        return False
    headers = (request.headers or {})
    value = headers.get("Sec-CH-Prefers-Color-Scheme") or headers.get("prefers-color-scheme")
    if not value:
        return False
    # The header is either a bare token ('dark') or a quality list ('dark;...').
    return value.split(",")[0].split(";")[0].strip().lower() == THEME_DARK


def resolve_theme(user):
    """Return ``(pref, resolved)`` for a user (or anonymous) without a request.

    For anonymous users there is no DB preference, so ``pref`` is empty and
    the inline script relies on localStorage / OS preference instead.
    """
    pref = ""
    if user is not None and getattr(user, "is_authenticated", False):
        pref = getattr(user, "theme", "") or THEME_SYSTEM
        if pref not in VALID_THEMES:
            pref = THEME_SYSTEM
    return pref, ""  # no request -> can't resolve 'system' server-side


def request_theme(request):
    """Return ``(pref, resolved)`` for a request, resolving 'system' if possible."""
    pref, _ = resolve_theme(getattr(request, "user", None))
    if pref == THEME_SYSTEM:
        resolved = THEME_DARK if _prefers_dark(request) else THEME_LIGHT
    else:
        resolved = pref or ""
    return pref, resolved


def theme_context(request):
    """Context processor: expose theme_pref + theme_resolved to all templates."""
    pref, resolved = request_theme(request)
    return {"theme_pref": pref, "theme_resolved": resolved}
